File: cpp/plugins/fixdictionarygenerator/src/FixDictionaryGenerator.cpp
#include "FixDictionaryGenerator.h"
#include <fstream>
#include <sstream>
#include <algorithm>

using namespace tinyxml2;

FixDictionaryGenerator::FixDictionaryGenerator() : document(new XMLDocument()) {}
FixDictionaryGenerator::~FixDictionaryGenerator() = default;

bool FixDictionaryGenerator::loadDictionary(const std::string& xmlFilePath) {
    if (document->LoadFile(xmlFilePath.c_str()) != XML_SUCCESS) {
        return false;
    }
    parseFields();
    parseMessages();
    parseComponents();
    return true;
}

void FixDictionaryGenerator::parseFields() {
    XMLElement* fieldsElem = document->FirstChildElement("fix")->FirstChildElement("fields");
    for (XMLElement* fieldElem = fieldsElem->FirstChildElement("field"); fieldElem; fieldElem = fieldElem->NextSiblingElement("field")) {
        FieldDefinition field;
        field.number = std::stoi(fieldElem->Attribute("number"));
        field.name = fieldElem->Attribute("name");
        field.type = fieldElem->Attribute("type");
        field.required = false;
        for (XMLElement* valueElem = fieldElem->FirstChildElement("value"); valueElem; valueElem = valueElem->NextSiblingElement("value")) {
            field.values[valueElem->Attribute("enum")] = valueElem->Attribute("description");
        }
        fields[field.name] = field;
    }
}

void FixDictionaryGenerator::parseMessages() {
    XMLElement* messagesElem = document->FirstChildElement("fix")->FirstChildElement("messages");
    for (XMLElement* msgElem = messagesElem->FirstChildElement("message"); msgElem; msgElem = msgElem->NextSiblingElement("message")) {
        MessageDefinition msg;
        msg.name = msgElem->Attribute("name");
        msg.msgType = msgElem->Attribute("msgtype");
        msg.msgCat = msgElem->Attribute("msgcat");
        for (XMLElement* fieldElem = msgElem->FirstChildElement("field"); fieldElem; fieldElem = fieldElem->NextSiblingElement("field")) {
            std::string fieldName = fieldElem->Attribute("name");
            bool required = std::string(fieldElem->Attribute("required")) == "Y";
            if (fields.count(fieldName)) {
                FieldDefinition field = fields[fieldName];
                field.required = required;
                msg.fields.push_back(field);
            }
        }
        messages[msg.msgType] = msg;
    }
}

void FixDictionaryGenerator::parseComponents() {
    XMLElement* compsElem = document->FirstChildElement("fix")->FirstChildElement("components");
    if (!compsElem) return;
    for (XMLElement* compElem = compsElem->FirstChildElement("component"); compElem; compElem = compElem->NextSiblingElement("component")) {
        ComponentDefinition comp;
        comp.name = compElem->Attribute("name");
        for (XMLElement* fieldElem = compElem->FirstChildElement("field"); fieldElem; fieldElem = fieldElem->NextSiblingElement("field")) {
            std::string fieldName = fieldElem->Attribute("name");
            bool required = std::string(fieldElem->Attribute("required")) == "Y";
            if (fields.count(fieldName)) {
                FieldDefinition field = fields[fieldName];
                field.required = required;
                comp.fields.push_back(field);
            }
        }
        components[comp.name] = comp;
    }
}

std::string FixDictionaryGenerator::toUpperCase(const std::string& str) {
    std::string result = str;
    std::transform(result.begin(), result.end(), result.begin(), ::toupper);
    return result;
}

std::string FixDictionaryGenerator::escapeString(const std::string& str) {
    std::string result = str;
    size_t pos = 0;
    while ((pos = result.find("\"", pos)) != std::string::npos) {
        result.replace(pos, 1, "\\\"");
        pos += 2;
    }
    return result;
}

std::string FixDictionaryGenerator::generateCppCode() {
    std::ostringstream code;
    code << "#pragma once\n#include <string>\n#include <map>\n#include <vector>\n\n";
    code << "class FixMessageCodec {\npublic:\n";
    code << "    // Field constants\n";
    for (const auto& [name, field] : fields) {
        code << "    static constexpr int " << toUpperCase(name) << " = " << field.number << ";\n";
    }
    code << "\n    // Message types\n";
    for (const auto& [msgType, msg] : messages) {
        code << "    static constexpr const char* " << toUpperCase(msg.name) << " = \"" << msgType << "\";\n";
    }
    code << "\n    static std::string encodeMessage(const std::string& msgType, const std::map<std::string, std::string>& fields);\n";
    code << "    static std::map<std::string, std::string> decodeMessage(const std::string& message);\n";
    code << "};\n";
    return code.str();
}

bool FixDictionaryGenerator::generateCodeToFile(const std::string& outputPath) {
    std::ofstream out(outputPath);
    if (!out) return false;
    out << generateCppCode();
    return true;
} 

File: python/plugins/fixdictionarygenerator/fixdictionarygenerator.py
import xml.etree.ElementTree as ET
from typing import Dict, List

class FieldDefinition:
    def __init__(self, number, name, type_):
        self.number = int(number)
        self.name = name
        self.type = type_
        self.values = {}
        self.required = False

class MessageDefinition:
    def __init__(self, name, msgtype, msgcat):
        self.name = name
        self.msgtype = msgtype
        self.msgcat = msgcat
        self.fields: List[FieldDefinition] = []

class FixDictionaryGenerator:
    def __init__(self):
        self.fields: Dict[str, FieldDefinition] = {}
        self.messages: Dict[str, MessageDefinition] = {}

    def load_dictionary(self, xml_file_path):
        tree = ET.parse(xml_file_path)
        root = tree.getroot()
        self._parse_fields(root)
        self._parse_messages(root)

    def _parse_fields(self, root):
        for field_elem in root.find('fields').findall('field'):
            field = FieldDefinition(
                field_elem.attrib['number'],
                field_elem.attrib['name'],
                field_elem.attrib['type']
            )
            for value_elem in field_elem.findall('value'):
                field.values[value_elem.attrib['enum']] = value_elem.attrib['description']
            self.fields[field.name] = field

    def _parse_messages(self, root):
        for msg_elem in root.find('messages').findall('message'):
            msg = MessageDefinition(
                msg_elem.attrib['name'],
                msg_elem.attrib['msgtype'],
                msg_elem.attrib['msgcat']
            )
            for field_elem in msg_elem.findall('field'):
                field_name = field_elem.attrib['name']
                required = field_elem.attrib['required'] == 'Y'
                if field_name in self.fields:
                    field = self.fields[field_name]
                    field.required = required
                    msg.fields.append(field)
            self.messages[msg.msgtype] = msg

    def generate_python_code(self):
        code = ["class FixMessageCodec:"]
        code.append("    # Field constants")
        for field in self.fields.values():
            code.append(f"    {field.name.upper()} = {field.number}")
        code.append("")
        code.append("    # Message types")
        for msg in self.messages.values():
            code.append(f"    {msg.name.upper()} = '{msg.msgtype}'")
        code.append("")
        code.append("    @staticmethod")
        code.append("    def encode_message(msg_type, fields):")
        code.append("        parts = [f'8=FIX.4.4\\x01', f'35={msg_type}\\x01']")
        code.append("        for k, v in fields.items():")
        code.append("            if v:")
        code.append("                parts.append(f'{FixMessageCodec.get_field_number(k)}={v}\\x01')")
        code.append("        message = ''.join(parts)")
        code.append("        body_length = len(message)")
        code.append("        message = f'8=FIX.4.4\\x019={body_length:09d}\\x01' + message[8:]")
        code.append("        checksum = sum(ord(c) for c in message) % 256")
        code.append("        message += f'10={checksum:03d}\\x01'")
        code.append("        return message")
        code.append("")
        code.append("    @staticmethod")
        code.append("    def decode_message(message):")
        code.append("        result = {}")
        code.append("        for part in message.split('\\x01'):")
        code.append("            if '=' in part:")
        code.append("                field, value = part.split('=', 1)")
        code.append("                result[FixMessageCodec.get_field_name(field)] = value")
        code.append("        return result")
        code.append("")
        code.append("    @staticmethod")
        code.append("    def get_field_number(field_name):")
        code.append("        mapping = {")
        for field in self.fields.values():
            code.append(f"            '{field.name}': '{field.number}',")
        code.append("        }")
        code.append("        return mapping.get(field_name, '0')")
        code.append("")
        code.append("    @staticmethod")
        code.append("    def get_field_name(field_number):")
        code.append("        mapping = {")
        for field in self.fields.values():
            code.append(f"            '{field.number}': '{field.name}',")
        code.append("        }")
        code.append("        return mapping.get(field_number, 'Unknown')")
        return '\n'.join(code)

    def generate_code_to_file(self, output_path):
        code = self.generate_python_code()
        with open(output_path, 'w') as f:
            f.write(code) 